File: C_plus_plus/Balanced_Binary_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode *root) {
        return dp(root).first;
    }
    pair<bool,int> dp(TreeNode *root){
        if(!root) return make_pair(true,0);
        pair<bool,int> a,b;
        a = dp(root->left);
        b = dp(root->right);
        if(!a.first||!b.first) return make_pair(false,0);
        if(abs(a.second-b.second)>1)
            return make_pair(false,0);
        return make_pair(true,(a.second>b.second?a.second:b.second)+1);
    }
};

File: java/Balanced_Binary_Tree.java
//  Compare each node's left and right subtree, if the two heights are not balanced, return -1,
//  or, return the node's height by formula max(left,right)+1.
//  Using this method, we calculate each node just once by dfs.
//  Time is O(n), space is O(n).

public class Solution {
    public boolean isBalanced(TreeNode root) {
        return getHeight(root) != -1;
    }
    public int getHeight(TreeNode root) {
        if (root == null) return 0;
        int left = getHeight(root.left);
        if (left == -1) return -1;
        int right = getHeight(root.right);
        if (right == -1) return -1;
        if (Math.abs(left-right) > 1) return -1;
        return Math.max(left,right)+1;
    }
}
